### PR TITLE
Added integration testing tools

### PR DESCRIPTION
**Pull request by lukaszb**
Please let me know if you're interested in those at all. What's wrong or missing?

### CODE REVIEW

**Comment by jsocol** [comment]
What's the use case here?

**Comment by lukaszb** [comment]
Let's imagine we have a part of the code that depends on multiple switches and want to test multiple scenarios. Usually I'm fine with mocking `waffle.switch_is_active` but when there are multiple switches I would need to _trap_ the mock and put some extra conditions there.

File: waffle/testutils.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        def test_happy_mode_enabled():
+            ...
+
+    """
+    def __init__(self, name, active):
+        self.name = name
+        self.active = active
+
+    def __call__(self, func):
+        def wrapped(*args, **kwargs):
```
Comment (jsocol) [inline waffle/testutils.py:31]: Can you add `@functools.wraps(func)` here?

File: waffle/testutils.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            with self:
+                return func(*args, **kwargs)
+        return wrapped
+
+    def update(self, active):
+        Switch.objects.filter(id=self.switch.id).update(active=active)
+
+    def __enter__(self):
+        self.switch, self.created = Switch.objects.get_or_create(name=self.name)
```
Comment (jsocol) [inline waffle/testutils.py:40]: Huh, I legitimately didn't know you could do that (have expressions on the LHS).

But it makes this line too long.

File: waffle/tests/test_testutils.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        assert not Switch.objects.get(name='foo').active
+
+    def test_new_switch(self):
+        with switched('foo', active=True):
+            assert waffle.switch_is_active('foo')
+
+        with switched('foo', active=False):
+            assert not waffle.switch_is_active('foo')
+
+        # make sure switch is removed (as it was created inside context manager)
```
Comment (jsocol) [inline waffle/tests/test_testutils.py:40]: This comment (and the one in the next test) is too long and if we replicated the assert before calling either method, I don't think they'd be necessary (i.e. this test assumes that the flag isn't there, so explicitly asserting that pre-condition then re-asserting it is fairly self-documenting).

**Comment by jsocol** [comment]
Ah OK, yeah, I see the value there. I'd like to pursue this but there are a couple of caveats.
- Anything done for one type needs to be done for the others, as long as it makes sense. So `switched` might not be the best name. Maybe following Django's `override_settings` example we should go with `override_{flag,sample,switch}`. It is also grammatically clearer.
- These need docs in the (horribly named) file `docs/testing-waffles.rst`.
- The inline comments I just made.

Totally understand if you're not up for adding similar code for flags and samples, but if not, I'll need to carve out some time to add it myself (or someone else can!) before any of this can make it into a released version.

**Comment by jsocol** [comment]
Thanks for the suggestion and starting code, @lukaszb! Added docs and flag/sample version and tests, and refactored a bit to support those.

**Comment by lukaszb** [comment]
@jsocol hey, thanks, sorry I couldn't help more. Too much going on recently (oh, I'm sure you know what I mean).

**Comment by jsocol** [comment]
I know very much what you mean. It's all good, and it was your idea! :+1:

**Pull request closed without merging**